Implement MissingDirChecker

zzhhhzz opened the pull request:
MissingDirChecker is a daemon thread that get initialized, started and stopped in LogStreamManager.  When LogStreamManager tries to create LogStreams for each SingerLog, if the corresponding log directory (eg: /mnt/thrift_logger/) has not been created yet, the SingerLog will be put to a HashMap named singerLogsWithoutDir. MissingDirChecker will run every sleep_in_milliseconds (eg: 20 seconds) to check if the log directory for this SingerLog has been created. Once log directory has been created, LogStreamManager.initializeLogStreams method will be called for this SingerLog and SingerLog  will be removed from singerLogsWithoutDir if the method call does not throw any exception.

zzhhhzz commented on singer/src/main/java/com/pinterest/singer/monitor/LogStreamManager.java:
```diff
@@ -96,7 +102,11 @@
   private RecursiveFSEventProcessor recursiveEventProcessor;
 
+  @VisibleForTesting
+  public MissingDirChecker missingDirChecker;
```
will change to getter & setter.

zzhhhzz commented on singer/src/main/java/com/pinterest/singer/monitor/LogStreamManager.java:
```diff
@@ -231,6 +253,10 @@ private void initializeLogStreamsInternal() throws SingerLogException {
           initializeLogStreamsInternal(NON_KUBERNETES_POD_ID, singerLog);
         }
+      LOG.info("set singerLogsWithoutDir and start MissingDirChecker thread.");
+      missingDirChecker.singerLogsWithoutDir = singerLogsWithoutDir;
```
will change to getter & setter.

zzhhhzz commented on singer/src/main/java/com/pinterest/singer/monitor/MissingDirChecker.java:
```diff
@@ -0,0 +1,99 @@
+  private static final Logger LOG = LoggerFactory.getLogger(MissingDirChecker.class);
+
+  @VisibleForTesting
+  public long sleep_in_milliseconds =  20000L;
```
will change to getter & setter.

zzhhhzz commented on singer/src/main/java/com/pinterest/singer/monitor/MissingDirChecker.java:
```diff
@@ -0,0 +1,99 @@
+  public long sleep_in_milliseconds =  20000L;
+
+  @VisibleForTesting
+  public boolean cancelled = false;
```
will change to AtomicBoolean

zzhhhzz commented on singer/src/test/java/com/pinterest/singer/monitor/MissingDirCheckerTest.java:
```diff
@@ -0,0 +1,131 @@
+  @After
+  public void tearDown() throws IOException {
+    File testBaseDir= tempDir.getRoot();
+    LOG.info("clean up files under test base dir: " + testBaseDir.getAbsolutePath());
```
Will make change to have consistent logging info.

zzhhhzz commented on singer/src/test/java/com/pinterest/singer/monitor/MissingDirCheckerTest.java:
```diff
@@ -0,0 +1,131 @@
+    LogStreamManager.initializeLogStreams();
+
+    // setting sleep time of MissingDirChecker to be 5 seconds in order to accelerate unit test.
+    instance.missingDirChecker.sleep_in_milliseconds= 5000;
```
will change to getter & setter.

zzhhhzz commented on singer/src/main/java/com/pinterest/singer/monitor/MissingDirChecker.java:
```diff
@@ -0,0 +1,99 @@
+  private static final Logger LOG = LoggerFactory.getLogger(MissingDirChecker.class);
+
+  @VisibleForTesting
+  public long sleep_in_milliseconds =  20000L;
```
will use camelCase.

zzhhhzz commented on singer/src/main/java/com/pinterest/singer/monitor/MissingDirChecker.java:
```diff
@@ -0,0 +1,99 @@
+    } catch (InterruptedException e){
+      LOG.warn("MissingDirChecker thread is interrupted ", e);
+    } finally {
+      LOG.info("MissingDirChecker thread stops. SingerLogs without dir: " + singerLogsWithoutDir);
```
will change to Stopped.

zzhhhzz commented on singer/src/test/java/com/pinterest/singer/monitor/MissingDirCheckerTest.java:
```diff
@@ -0,0 +1,131 @@
+      File f1 = new File(path);
+      f1.mkdirs();
+      assertTrue(f1.exists() && f1.isDirectory());
+      Thread.sleep(instance.missingDirChecker.sleep_in_milliseconds * 2);
```
Good point! I  will add a note.

zzhhhzz commented on singer/src/test/java/com/pinterest/singer/monitor/MissingDirCheckerTest.java:
```diff
@@ -0,0 +1,131 @@
+
+    LOG.info("use java reflection to make sure LogStreamManager instance (singleton) is "
+        + "specifically set for this unit test");
+    Field instance = LogStreamManager.class.getDeclaredField("instance");
```
use get and reset to make sure the LogStreamManager instance is specifically set for this unit test.  will no longer use java reflection.

zzhhhzz commented on singer/src/main/java/com/pinterest/singer/monitor/LogStreamManager.java:
```diff
@@ -231,6 +253,10 @@ private void initializeLogStreamsInternal() throws SingerLogException {
         }
+      LOG.info("set singerLogsWithoutDir and start MissingDirChecker thread.");
+      missingDirChecker.singerLogsWithoutDir = singerLogsWithoutDir;
+      missingDirChecker.start();
```
MissingDirChecker only track SingerLog with NON_KUBERNETES_POD_ID (eg: SingerLogs as non kubenetes pods).  There should be no conflict with KubeService and RecursiveFSEventProcessor.

zzhhhzz commented on singer/src/main/java/com/pinterest/singer/monitor/MissingDirChecker.java:
```diff
@@ -0,0 +1,99 @@
+          if (dir.exists()) {
+            try {
+              LogStreamManager.initializeLogStreams(singerLog, podUid);
+              iterator.remove();
```
metrics (SingerMetrics.NUMBER_OF_MISSING_DIRS) added.

zzhhhzz commented on singer/src/main/java/com/pinterest/singer/monitor/MissingDirChecker.java:
```diff
@@ -0,0 +1,99 @@
+        Thread.sleep(sleep_in_milliseconds);
+      }
+    } catch (InterruptedException e){
+      LOG.warn("MissingDirChecker thread is interrupted ", e);
```
metric (SingerMetrics.MISSING_DIR_CHECKER_INTERRUPTED) added.

ambud commented on singer/src/main/java/com/pinterest/singer/monitor/MissingDirChecker.java:
```diff
@@ -0,0 +1,138 @@
+    } catch (InterruptedException e){
+      Stats.incr(SingerMetrics.MISSING_DIR_CHECKER_INTERRUPTED);
+      LOG.warn("MissingDirChecker thread is interrupted ", e);
+    } finally {
```
We should add a catch statement to log exception if the Exception is other than InterruptedException in case there is an issue `finally` block will now log what happened and the thread will simply get terminated.

zzhhhzz replied:
added

ambud commented on singer/src/main/java/com/pinterest/singer/monitor/MissingDirChecker.java:
```diff
@@ -0,0 +1,138 @@
+    }
+  }
+
+  public void start() {
```
add `synchronized` ? (to prevent potential duplicate start)

zzhhhzz replied:
added in update.

ambud commented on singer/src/main/java/com/pinterest/singer/monitor/MissingDirChecker.java:
```diff
@@ -0,0 +1,138 @@
+    }
+  }
+
+  public void stop(){
```
add `synchronized` ?

zzhhhzz replied:
added

ambud commented on singer/src/main/java/com/pinterest/singer/monitor/MissingDirChecker.java:
```diff
@@ -0,0 +1,138 @@
+    } catch (InterruptedException e){
+      Stats.incr(SingerMetrics.MISSING_DIR_CHECKER_INTERRUPTED);
+      LOG.warn("MissingDirChecker thread is interrupted ", e);
+    } finally {
```
We should add a catch-all to log the reason of thread termination else with a finally clause we wouldn't know why the thread stopped.

zzhhhzz replied:
added

ambud commented on singer/src/main/java/com/pinterest/singer/monitor/MissingDirChecker.java:
```diff
@@ -0,0 +1,138 @@
+    }
+  }
+
+  public void start() {
```
Make method synchronized to prevent concurrent initialization issue since there is a conditional check.

yuyang08 commented on singer/src/main/java/com/pinterest/singer/monitor/MissingDirChecker.java:
```diff
@@ -0,0 +1,140 @@
+        Thread.currentThread().getName()));
+    try {
+      while (!cancelled.get() && singerLogsWithoutDir != null && singerLogsWithoutDir.size() >0) {
+        LOG.info(String.format("[%s] found %d singerLogs without directory",
```
Simplify the line as follows?   `String.format` is optional for LOG.info

          LOG.info("[{}] found {} singerLogs without directory",
              Thread.currentThread().getName(), singerLogsWithoutDir.size());

yuyang08 commented on singer/src/main/java/com/pinterest/singer/monitor/MissingDirChecker.java:
```diff
@@ -0,0 +1,140 @@
+      while (!cancelled.get() && singerLogsWithoutDir != null && singerLogsWithoutDir.size() >0) {
+        LOG.info(String.format("[%s] found %d singerLogs without directory",
+            Thread.currentThread().getName(), singerLogsWithoutDir.size()));
+        Iterator<Map.Entry<SingerLog, String>>
```
merge line 85 -87 as one line?

yuyang08 commented on singer/src/main/java/com/pinterest/singer/monitor/MissingDirChecker.java:
```diff
@@ -0,0 +1,140 @@
+              LogStreamManager.initializeLogStreams(singerLog, podUid);
+              iterator.remove();
+              Stats.setGauge(SingerMetrics.NUMBER_OF_MISSING_DIRS,  singerLogsWithoutDir.size());
+              LOG.info(String.format("[%s] after %s created, log stream is initialized for %s.",
```
same here.

              LOG.info("[{}] after %s created, log stream is initialized for {}.",
                  Thread.currentThread().getName(), dir.getAbsoluteFile(), singerLog.getLogName());

yuyang08 commented on singer/src/main/java/com/pinterest/singer/monitor/MissingDirChecker.java:
```diff
@@ -0,0 +1,140 @@
+    return singerLogsWithoutDir;
+  }
+
+  public void setSingerLogsWithoutDir(
```
merge line 72-73 as one line?

yuyang08 commented on singer/src/main/java/com/pinterest/singer/monitor/MissingDirChecker.java:
```diff
@@ -0,0 +1,140 @@
+
+  private static final Logger LOG = LoggerFactory.getLogger(MissingDirChecker.class);
+
+  private long sleepInMills =  20000L;
```
format the code to remove extra space before `20000L`?

zzhhhzz commented:
@yuyang08  Thank you for the review. I updated the PR.   In terms of you high level question,  kubernetes pod are created and deleted dynamically and the directory checking has been handled by RecursiveFSEventProcessor.java and  KubeService.java     The MissingDirChecker  added in the PR is to handle NON kubernetes POD use case.
